library/sha3.c

```diff
@@ -26,25 +26,6 @@
 
 #define XOR_BYTE 0x6
 
-typedef struct mbedtls_sha3_family_functions {
-    mbedtls_sha3_id id;
-
-    uint16_t r;
-    uint16_t olen;
-}
-mbedtls_sha3_family_functions;
-
-/*
- * List of supported SHA-3 families
- */
-static const mbedtls_sha3_family_functions sha3_families[] = {
-    { MBEDTLS_SHA3_224,      1152, 224 },
-    { MBEDTLS_SHA3_256,      1088, 256 },
-    { MBEDTLS_SHA3_384,       832, 384 },
-    { MBEDTLS_SHA3_512,       576, 512 },
-    { MBEDTLS_SHA3_NONE, 0, 0 }
-};
-
 static const uint64_t rc[24] = {
```

Comment:
Not really related to your changes, but this table is pretty sparse. We can save an extra 128 bytes by encoding it as:

```
static const uint32_t rc63 = 0xbbe0cc; // pack bit 63 of rc table
static const uint32_t rc31 = 0xd81c68; // pack bit 31 of rc table
static const uint16_t rc[24] = {
    0x0001, 0x8082, 0x808a, 0x8000,
    0x808b, 0x0001, 0x8081, 0x8009,
    0x008a, 0x0088, 0x8009, 0x000a,
    0x808b, 0x008b, 0x8089, 0x8003,
    0x8002, 0x0080, 0x800a, 0x000a,
    0x8081, 0x8080, 0x0001, 0x8008,
};
#define RC(i) (rc[i] | (((rc63 >> round) & 1ull) << 63) | (((rc31 >> round) & 1ull) << 31))
```

(please go ahead & add this to this PR if you think it's a useful saving)

Reply:
I played with this (https://github.com/gilles-peskine-arm/mbedtls/tree/sha3-no-table%2Bsparse-rc-experiment) and while this saves code size, it also has a measurable performance impact on my PC. Even just isolating `rc63`, which is half of the code size saving, seems to have a performance cost, although I'm struggling a bit to get reliable numbers.

Reply:
Interesting, I made a very very quick test (M1 pro) and saw ~2% perf regression with the 16-bit table, which I'd think is a reasonable price to pay for the saving. My guess is that embedded might get some benefit from reducing memory bandwidth in the absence of a cache, but I haven't actually tested this.

I didn't try the 8-bit table (`RC_SIZE == 1`) as I guessed that the 24 byte additional saving would be largely offset by the extra code size.

Reply:
Testing it slightly more carefully now (best of 10 runs of SHA3-256 with clang -Os on M1 Pro), I get:

RC_SIZE 4: 151701
RC_SIZE 3: 150733
RC_SIZE 2: 151646
RC_SIZE 1: 148383

Reply:
gcc shows a slightly bigger variation for me, but not a lot:
RC_SIZE 4: 155194
RC_SIZE 3: 150229
RC_SIZE 2: 149990
RC_SIZE 1: 149369

i.e. < 4% difference

Reply:
Even on par with RC_SIZE=1. Can you please try `RC_SIZE=5` and `RC_SIZE=6` on your machine? They're both performance improvements on my PC (no visible difference between them), and 6 is more compact than the others.

Reply:
In any case, given that there are two approvals here and it's out of scope, I'll make it a follow-up PR.

Reply:
> I'm getting about 7% increase in performance on my PC

That was with `gcc -O3`. With `clang -O3` (still x86_64), all the variants with a significant code size saving have a ~5% penalty. Still worth it, I think. Especially since Clang's baseline is about 2× faster than GCC!

Reply:
I think < 5% for a reasonable size win is a good trade-off.